src/vs/workbench/contrib/terminal/browser/xterm/xtermTerminal.ts

```diff
@@ -763,76 +771,119 @@ export class XtermTerminal extends DisposableStore implements IXtermTerminal, II
+	}
+
+	private _getShellIntegrationContent(): string {
 		const commands = this._capabilities.get(TerminalCapability.CommandDetection)?.commands;
-		if (!commands?.length) {
-			const noContent = document.createElement('div');
-			const noContentLabel = localize('terminal.integrated.noContent', "No terminal content available for this session.");
-			noContent.textContent = noContentLabel;
-			this._bufferElementFragment.replaceChildren(noContent);
-			this._accessibleBuffer.focus();
-			return this._bufferElementFragment;
-		}
-		let header;
-		let replaceChildren = true;
-		for (const command of commands) {
-			header = document.createElement('h2');
-			// without this, the text area gets focused when keyboard shortcuts are used
-			header.tabIndex = -1;
-			header.textContent = command.command.replace(new RegExp(' ', 'g'), '\xA0');
-			if (command.exitCode !== 0) {
-				header.textContent += ` exited with code ${command.exitCode}`;
+		const sb = new StringBuilder(10000);
+		let content = localize('terminal.integrated.noContent', "No terminal content available for this session. Run some commands to create content.");
+		if (commands?.length) {
+			for (const command of commands) {
+				sb.appendString(command.command.replace(new RegExp(' ', 'g'), '\xA0'));
+				if (command.exitCode !== 0) {
+					sb.appendString(` exited with code ${command.exitCode}`);
+				}
+				sb.appendString('\n');
+				sb.appendString(command.getOutput()?.replace(new RegExp(' ', 'g'), '\xA0') || '');
 			}
-			const output = document.createElement('div');
-			// without this, the text area gets focused when keyboard shortcuts are used
-			output.tabIndex = -1;
-			output.textContent = command.getOutput()?.replace(new RegExp(' ', 'g'), '\xA0') || '';
-			if (replaceChildren) {
-				this._bufferElementFragment.replaceChildren(header, output);
-				replaceChildren = false;
-			} else {
-				this._bufferElementFragment.appendChild(header);
-				this._bufferElementFragment.appendChild(output);
+			content = sb.build();
+		}
+		return content;
+	}
+
+	private _getAllContent(): string {
+		let content = '';
+		let currentLine: string = '';
+		const end = this._terminal.buffer.active.length;
+		for (let i = 0; i < end; i++) {
+			const line = this._terminal.buffer.active.getLine(i);
+			if (!line) {
+				continue;
+			}
+			const isWrapped = this._terminal.buffer.active.getLine(i + 1)?.isWrapped;
+			currentLine += line.translateToString(!isWrapped);
+			if (!isWrapped || i === end - 1) {
+				content += currentLine.replace(new RegExp(' ', 'g'), '\xA0');
+				currentLine = '';
 			}
 		}
-		this._accessibleBuffer.focus();
-		if (this._accessibleBuffer.contentEditable === 'true') {
-			document.execCommand('selectAll', false, undefined);
-			document.getSelection()?.collapseToEnd();
-		} else if (header) {
-			// focus the cursor line's header
-			header.tabIndex = 0;
-		}
-		return this._bufferElementFragment;
+		return content;
```

Comment:
It'll save a bunch of memory being allocated if you do the `lines: string[] = []`/`lines.join('\n')` approach (or stringbuilder again?)